doc/barclamp.md

```diff
@@ -0,0 +1,1321 @@
+# fork, though. It also assumes that you are not working on your fork's
+# `master` branch (you shouldn't, but people have been known to do this
+# anyway). Since all of these may occasionally happen, this guide leaves it up
+# to the user to set $COMMIT as they see fit rather than trying to determine
+# its value automatically through the command mentioned above.
+
+export COMMIT="6cefbd80254141a63f4c86df195cd930ac7eb10b"
+
+# Insert your own barclamp's name here:
+export BARCLAMP_NAME=mybarclamp
+
+# Insert a space delimited list of the Barclamps you are working on here (only
+# required if your barclamp requires changes to other barclamps as well).
+
+export BARCLAMPS="$BARCLAMP_NAME"
+
+# Insert the path to your barclamp collection checkout here (in case it's
+# somewhere other than /root/crowbar-openstack). This directory must be a git
+# working copy of the crowbar-openstack fork and branch you are working on.
+
+export BARCLAMP_FORK_PATH=/root/crowbar-openstack
+
+export GIT_AUTHOR_NAME="$C_NAME"
+export GIT_AUTHOR_EMAIL="$C_EMAIL"
+export GIT_COMMITTER_NAME="$C_NAME"
+export GIT_COMMITTER_EMAIL="$C_EMAIL"
+
+# Applies all the committed changes in your crowbar-openstack working copy to
+# /opt/dell and updates Crowbar runtime state. Note: this will break if you
+# rebase your working copy, so do not rebase your working copy. If you rebase
+# it (e.g. for final testing before a merge, please build a fresh Crowbar node
+# or restore it from a snapshot).
+function reposync_crowbar()
+  {
+  if [ -e /opt/dell/patchlevel ]; then
+    . /opt/dell/patchlevel
+    head=$(cd $BARCLAMP_FORK_PATH; git log --format='%H'| head -n 1)
+    if [ "$head" = "$commit" ]; then
+      echo "/opt/dell is already up to date. If this is not what you " 1>&2
+      echo "expected to see, you may have uncommitted changes" 1>&2
+      echo "in ${BARCLAMP_FORK_PATH}." 1>&2
+      return
+    fi
+  else
+    commit="${COMMIT}^"
+  fi
+
+  pushd $BARCLAMP_FORK_PATH > /dev/null
+  cp -a /opt/dell /opt/dell.orig
+  git diff $commit HEAD | patch -N --merge -p1 -d /opt/dell || return
+  rm -rf /opt/dell.orig
+  # Record current HEAD as patchlevel
+  echo "commit=$(git log --format='%H'| head -n 1)" > /opt/dell/patchlevel
+  popd > /dev/null
+
+  for barclamp in $BARCLAMPS
+    do
+    if [ ! -h /opt/dell/crowbar_framework/barclamps/${barclamp}.yml ]; then
+      ln -s /opt/dell/${barclamp}.yml /opt/dell/crowbar_framework/barclamps 2>&1 > /dev/null
+    fi
+    done
+  chmod 755 /opt/dell/bin/*
+
+  sync_crowbar
+  }
+
+# Helper function: updates Crowbar's runtime state after /opt/dell
+# has been modified.
+sync_crowbar()
+  {
+  barclamp_install.rb /opt/dell/crowbar_framework/barclamps &&
+  for barclamp in $BARCLAMPS
+    do
+    knife cookbook upload -o /opt/dell/chef/cookbooks/ $barclamp
+    done
+  systemctl restart crowbar
+  }
+
+```
+
+We will make heavy use of the `reposync_crowbar` command thus defined in the
```

Comment:
> Please can you take a look at https://github.com/crowbar/crowbar/blob/master/Guardfile and
http://publishwith.me/crowbar-tree-merging and then consider whether a more unified approach might be possible? (I'm not claiming it would be easy, but at least it would tackle the fundamental problem that the git repo structure doesn't match the deployed structure.)

Not everybody will happy to perform
https://github.com/crowbar/crowbar/blob/master/Guardfile#L20 on their local
machine. But, more importantly, this approach is awkward in a collaboration scenario. That being said I added a link for those who prefer
working locally.

As for http://publishwith.me/crowbar-tree-merging: I'm cautiously optimistic
that we might have a unified solution at some stage, but the reality is that
we do not have a unified solution right now. Nor do I intend to provide one
with that little hack. It's just something a new developer can use for their
first steps (after that they'll build something of their own anyway, and _that_
may end up they unified solution we are looking for).

> If you'll forgive the history lesson, a long time ago in a universe far away, someone started to write a shell script a bit like this, with the goal of making Crowbar development smoother. It grew into an 8,000-line monstrosity which could have been avoided by asking some more fundamental questions about development environments and workflows. I'm not saying there's a similar risk here, but I think the KISS principle still applies :) Would be very interested in your thoughts.

I bore the KISS principle in mind as I came up with this :-)

While the script grew some tentacles, still I tried to limit these as far as
possible. Also, I put it in here on purpose: I consider it more
documentation/inspiration than code to be run as-is since it shows what needs
to be done to synchronize the state of one's `crowbar-openstack` fork with a
running Crowbar instance. I only intend to change it if these synchronization
needs change. It is definitely not supposed to solve everybody's development
environment problems. Where that lofty goal is concerned its only aspiration is
showing people what challenges they face if they chose to work on that solution.